File: news/fetch_and_notify.py
"""
AI Daily LINE Notifier — Tier 2
รันผ่าน Hostinger Cron Job หรือ GitHub Actions
"""
import sys
import json
import logging
import requests
from datetime import datetime, timezone, timedelta
from pathlib import Path
from dotenv import load_dotenv
import os

# --- Config ---
BASE_DIR = Path(__file__).resolve().parent.parent
load_dotenv(BASE_DIR / ".env")

LOG_FILE = BASE_DIR / "logs" / "daily_news.log"
LOG_FILE.parent.mkdir(exist_ok=True)

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s [%(levelname)s] %(message)s",
    handlers=[
        logging.FileHandler(LOG_FILE, encoding="utf-8"),
        logging.StreamHandler(sys.stdout),
    ],
)
log = logging.getLogger(__name__)

GNEWS_API_KEY          = os.environ["GNEWS_API_KEY"]
GEMINI_API_KEY         = os.environ["GEMINI_API_KEY"]
LINE_TOKEN             = os.environ["LINE_CHANNEL_ACCESS_TOKEN"]
LINE_USER_ID           = os.environ["LINE_USER_ID"]

GNEWS_URL = "https://gnews.io/api/v4/search"
GNEWS_KEYWORDS = "AI OR Generative AI OR Machine Learning OR Tech Innovation"

ICT = timezone(timedelta(hours=7))
TODAY = datetime.now(ICT).strftime("%d/%m/%Y")

SYSTEM_PROMPT = f"""คุณคือบรรณาธิการข่าว AI ระดับมืออาชีพ สรุปข่าวเป็นภาษาไทยตาม Template นี้เท่านั้น:

🤖 AI DAILY | {TODAY}
ธีมวันนี้: [ธีมสั้นๆ 1 ประโยค]

【บทสรุป】
[สรุปภาพรวมของวัน 2-3 ประโยค]

🇹🇭 ผลกระทบต่อไทย
[วิเคราะห์ผลกระทบเชิงธุรกิจและสังคมไทย 2-3 ประโยค]

【ข่าวเด่น】
① [หัวข้อ]
[สรุป 1-2 ประโยค] → So what: [ผลกระทบ]

② [หัวข้อ]
[สรุป 1-2 ประโยค] → So what: [ผลกระทบ]

③ [หัวข้อ]
[สรุป 1-2 ประโยค] → So what: [ผลกระทบ]

👀 จับตา
[ประเด็นที่ต้องติดตามต่อ 1-2 ประโยค]

📎 ที่มา
[ชื่อข่าว 1]: [URL]
[ชื่อข่าว 2]: [URL]
[ชื่อข่าว 3]: [URL]

กฎเหล็ก: ห้ามเพิ่มหัวข้อนอก Template, ใช้ภาษาไทยทั้งหมด, รักษาโครงสร้างให้ครบถ้วน"""


def fetch_news() -> list[dict]:
    """ดึงข่าวจาก GNews API"""
    log.info("Fetching news from GNews API...")
    resp = requests.get(GNEWS_URL, params={
        "q": GNEWS_KEYWORDS,
        "token": GNEWS_API_KEY,
        "lang": "en",
        "max": 10,
        "sortby": "publishedAt",
    }, timeout=15)
    resp.raise_for_status()
    articles = resp.json().get("articles", [])
    log.info(f"Fetched {len(articles)} articles")
    return articles


def summarize_with_gemini(articles: list[dict]) -> str:
    """สรุปข่าวด้วย Gemini 1.5 Flash"""
    log.info("Summarizing with Gemini 1.5 Flash...")
    news_text = "\n\n".join([
        f"Title: {a['title']}\nSource: {a['source']['name']}\nURL: {a['url']}\nContent: {a.get('description', '')}"
        for a in articles[:10]
    ])
    prompt = f"{SYSTEM_PROMPT}\n\nข่าววันนี้:\n\n{news_text}"
    url = f"https://generativelanguage.googleapis.com/v1/models/gemini-1.5-flash:generateContent?key={GEMINI_API_KEY}"
    resp = requests.post(url, json={
        "contents": [{"parts": [{"text": prompt}]}],
        "generationConfig": {"maxOutputTokens": 1500, "temperature": 0.4},
    }, timeout=30)
    resp.raise_for_status()
    summary = resp.json()["candidates"][0]["content"]["parts"][0]["text"].strip()
    log.info(f"Summary generated ({len(summary)} chars)")
    return summary


def send_line(message: str) -> None:
    """ส่งข้อความไปยัง LINE Messaging API"""
    log.info("Sending to LINE...")
    resp = requests.post(
        "https://api.line.me/v2/bot/message/push",
        headers={
            "Authorization": f"Bearer {LINE_TOKEN}",
            "Content-Type": "application/json",
        },
        json={"to": LINE_USER_ID, "messages": [{"type": "text", "text": message}]},
        timeout=15,
    )
    resp.raise_for_status()
    log.info(f"LINE sent OK — message ID: {resp.json()['sentMessages'][0]['id']}")


def main():
    log.info("=== AI Daily LINE Notifier started ===")
    try:
        articles = fetch_news()
        if not articles:
            log.warning("No articles found — skipping")
            return
        summary = summarize_with_gemini(articles)
        send_line(summary)
        log.info("=== Done ===")
    except requests.HTTPError as e:
        log.error(f"HTTP error: {e.response.status_code} {e.response.text}")
        sys.exit(1)
    except Exception as e:
        log.error(f"Unexpected error: {e}", exc_info=True)
        sys.exit(1)


if __name__ == "__main__":
    main()

File: fetch_and_notify.js
/**
 * AI Daily LINE Notifier — Tier 2 (Node.js)
 * รันผ่าน Hostinger Node.js Web App + Cron
 */

const https = require("https");
const http = require("http");
const fs = require("fs");
const path = require("path");
const express = require("express");
const cron = require("node-cron");
require("dotenv").config({ path: path.join(__dirname, ".env") });

// --- Config ---
const GNEWS_API_KEY = process.env.GNEWS_API_KEY;
const GEMINI_API_KEY = process.env.GEMINI_API_KEY;
const LINE_TOKEN = process.env.LINE_CHANNEL_ACCESS_TOKEN;
const LINE_USER_ID = process.env.LINE_USER_ID;

const GNEWS_KEYWORDS = "AI OR Generative AI OR Machine Learning OR Tech Innovation";
const ICT_OFFSET = 7 * 3600 * 1000;
const THAI_MONTHS = ["มกราคม","กุมภาพันธ์","มีนาคม","เมษายน","พฤษภาคม","มิถุนายน",
  "กรกฎาคม","สิงหาคม","กันยายน","ตุลาคม","พฤศจิกายน","ธันวาคม"];
const _now = new Date(Date.now() + ICT_OFFSET);
const TODAY = `${_now.getUTCDate()} ${THAI_MONTHS[_now.getUTCMonth()]} ${_now.getUTCFullYear() + 543}`;

const LOG_DIR = path.join(__dirname, "logs");
const LOG_FILE = path.join(LOG_DIR, "daily_news.log");

// Ensure log dir exists
if (!fs.existsSync(LOG_DIR)) fs.mkdirSync(LOG_DIR, { recursive: true });

function log(level, msg) {
  const timestamp = new Date().toISOString();
  const line = `${timestamp} [${level}] ${msg}`;
  console.log(line);
  fs.appendFileSync(LOG_FILE, line + "\n", "utf-8");
}

function httpGet(url, headers = {}) {
  return new Promise((resolve, reject) => {
    const protocol = url.startsWith("https") ? https : http;
    protocol
      .get(url, { headers }, (res) => {
        let data = "";
        res.on("data", (chunk) => (data += chunk));
        res.on("end", () => {
          if (res.statusCode >= 200 && res.statusCode < 300) {
            resolve(JSON.parse(data));
          } else {
            reject(new Error(`HTTP ${res.statusCode}: ${data}`));
          }
        });
      })
      .on("error", reject);
  });
}

function httpPost(url, data, headers = {}) {
  return new Promise((resolve, reject) => {
    const body = JSON.stringify(data);
    const options = {
      method: "POST",
      headers: { "Content-Type": "application/json", ...headers },
    };
    const protocol = url.startsWith("https") ? https : http;
    const req = protocol.request(url, options, (res) => {
      let responseData = "";
      res.on("data", (chunk) => (responseData += chunk));
      res.on("end", () => {
        if (res.statusCode >= 200 && res.statusCode < 300) {
          resolve(JSON.parse(responseData));
        } else {
          reject(new Error(`HTTP ${res.statusCode}: ${responseData}`));
        }
      });
    });
    req.on("error", reject);
    req.write(body);
    req.end();
  });
}

async function fetchNews() {
  log("INFO", "Fetching news from GNews API...");
  const url =
    "https://gnews.io/api/v4/search?" +
    new URLSearchParams({
      q: GNEWS_KEYWORDS,
      token: GNEWS_API_KEY,
      lang: "en",
      max: 10,
      sortby: "publishedAt",
    });
  const data = await httpGet(url);
  log("INFO", `Fetched ${data.articles.length} articles`);
  return data.articles;
}

async function summarizeWithGemini(articles) {
  log("INFO", "Summarizing with Gemini 1.5 Flash...");

  const newsText = articles
    .slice(0, 10)
    .map(
      (a) =>
        `Title: ${a.title}\nSource: ${a.source.name}\nURL: ${a.url}\nContent: ${a.description || ""}`
    )
    .join("\n\n");

  const prompt = `คุณคือบรรณาธิการข่าว AI สรุปข่าวเป็นภาษาไทยตาม Template นี้เท่านั้น ห้ามเพิ่มหรือลดหัวข้อ:

🤖 AI DAILY | ${TODAY}
ธีมวันนี้: [คำเดียวหรือวลีสั้น]

【บทสรุป】
• [ประเด็นสำคัญที่ 1]
• [ประเด็นสำคัญที่ 2]
• [ประเด็นสำคัญที่ 3]

🇹🇭 ผลกระทบต่อไทย
• [ใคร/อะไร/ต้องทำอะไร — เลือก 1 ข้อสำคัญที่สุด]

【ข่าวเด่น】
1) [หัวข้อข่าว]
   → [so-what 1 บรรทัด]
2) [หัวข้อข่าว]
   → [so-what 1 บรรทัด]
3) [หัวข้อข่าว]
   → [so-what 1 บรรทัด]

👀 จับตา
• [เหตุการณ์ที่น่าจับตา + กรอบเวลา]

📎 ที่มา
- [ชื่อสำนักข่าวของข่าวที่ 1] · [URL จริงของข่าวที่ 1]
- [ชื่อสำนักข่าวของข่าวที่ 2] · [URL จริงของข่าวที่ 2]
- [ชื่อสำนักข่าวของข่าวที่ 3] · [URL จริงของข่าวที่ 3]

กฎเหล็ก: ใช้ภาษาไทยทั้งหมด, ห้ามเพิ่มหัวข้อนอก Template, แต่ละ bullet กระชับที่สุด, ใช้ชื่อสำนักข่าวและ URL จากข้อมูลที่ให้มาเท่านั้น ห้ามแต่ง URL ขึ้นมาเอง

ข่าววันนี้:

${newsText}`;

  const response = await httpPost(
    `https://generativelanguage.googleapis.com/v1beta/models/gemini-2.5-flash:generateContent?key=${GEMINI_API_KEY}`,
    {
      contents: [{ parts: [{ text: prompt }] }],
      generationConfig: { maxOutputTokens: 2048, temperature: 0.4 },
    }
  );

  const summary = response.candidates[0].content.parts[0].text.trim();
  log("INFO", `Summary generated (${summary.length} chars)`);
  return summary;
}

async function sendLine(message) {
  log("INFO", "Sending to LINE...");
  const response = await httpPost(
    "https://api.line.me/v2/bot/message/broadcast",
    {
      messages: [{ type: "text", text: message }],
    },
    { Authorization: `Bearer ${LINE_TOKEN}` }
  );
  log(
    "INFO",
    `LINE broadcast OK — message ID: ${response.sentMessages[0].id}`
  );
}

async function main() {
  log("INFO", "=== AI Daily LINE Notifier started ===");
  const articles = await fetchNews();
  if (!articles || articles.length === 0) {
    log("WARN", "No articles found — skipping");
    return;
  }
  const summary = await summarizeWithGemini(articles);
  await sendLine(summary);
  log("INFO", "=== Done ===");
}

// --- Express Server ---
const app = express();
const PORT = process.env.PORT || 3000;
const CRON_SECRET = process.env.CRON_SECRET || "";

app.get("/", (_req, res) => res.send("AI Daily Notifier is running ✅"));

app.get("/run", async (req, res) => {
  if (CRON_SECRET && req.query.token !== CRON_SECRET) {
    return res.status(401).send("Unauthorized");
  }
  try {
    await main();
    res.send("Done ✅");
  } catch (err) {
    log("ERROR", err.message);
    res.status(500).send(`Error: ${err.message}`);
  }
});

app.listen(PORT, () => log("INFO", `Server listening on port ${PORT}`));

// --- Cron Job: รันทุกวัน 08:00 ICT (01:00 UTC) ---
cron.schedule("0 8 * * *", () => {
  log("INFO", "Cron triggered — running daily job...");
  main().catch((err) => log("ERROR", err.message));
}, { timezone: "Asia/Bangkok" });
log("INFO", "Cron registered: daily at 08:00 Asia/Bangkok");
